pip/pip-282.md

```diff
@@ -0,0 +1,293 @@
+   c2: messagesForC: 1000, pending: []  
+   c3: messagesForC: 998, pending: [`2:2`, `2:1`]  // out-of-order  
+   selector: key-a: c3
+
+# Goals
+
+## In Scope
+
+Fix out-of-order issues above.
+
+## Out of Scope
+
+Simplify or improve the specification of Key_Shared.
+
+# High Level Design
+
+The root cause of the issues described above is that `recentlyJoinedConsumers` uses "read position" as joined positions for consumers, because this does not guarantee that messages less than or equal to it have already been scheduled to be sent.
+Instead, we propose to use "last sent position" as joined positions for consumers.
+
+Also, change (or add) some stats to know Key_Shared subscription status easily.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+First, introduce the new position, like the mark delete position and the individually deleted messages. In other words,
+
+- All positions less than or equal to it are already scheduled to be sent.
+- Manage individually sent positions to update the position as expected.
+
+An example of updating the individually sent messages and the last sent position will be as follows.
+
+Initially, the last sent position is `3:0`, and the individually sent positions is `[]`.
+1. Read `3:1` - `3:10` positions
+2. Send `3:1` - `3:3`, `3:5`, and `3:8` - `3:10` positions
+   - last sent position: `3:3`
+   - individually sent positions: `[(3:4, 3:5], (3:7, 3:10]]`
+3. Send `3:7` position
+   - last sent position: `3:3`
+   - individually sent positions: `[(3:4, 3:5], (3:6, 3:10]]`
+4. Send `3:6` position
+   - last sent position: `3:3`
+   - individually sent positions: `[(3:4, 3:10]]`
+5. Send `3:4` position
+   - last sent position: `3:10`
+   - individually sent positions: `[]`
+
+More specifically, the recently joined consumers related fields will be as follows.
+```diff
+diff --git a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java
+index 8f05530f58b..2b17c580832 100644
+--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java
++++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java
+@@ -69,8 +69,12 @@ public class PersistentStickyKeyDispatcherMultipleConsumers extends PersistentDi
+      * This means that, in order to preserve ordering, new consumers can only receive old
+      * messages, until the mark-delete position will move past this point.
+      */
++    // Map(key: recently joined consumer, value: last sent position when joining)
+     private final LinkedHashMap<Consumer, PositionImpl> recentlyJoinedConsumers;
+
++    private PositionImpl lastSentPosition;
++    private final RangeSetWrapper<PositionImpl> individuallySentPositions;
```

Comment:
- `individuallySentPositions` = `{Range-Full} - {cursor.individualDeletedMessages} - {dispatcher.redeliveryMessages} - {the positions in inflight Replay Reading}`, right?
- Should we add metrics to describe how much memory `individuallySentPositions` usage? 
- Should we add a mechanism to stop delivering messages to the client if `individuallySentPositions` uses more memory than expected?

Reply:
@equanz Sorry for the late reply. Could you take a look at these questions?

Reply:
> individuallySentPositions = {Range-Full} - {cursor.individualDeletedMessages} - {dispatcher.redeliveryMessages} - {the positions in inflight Replay Reading}, right?

Roughly, the `{cursor.individualDeletedMessages}` is not subtracted. It is initialized by `individualDeletedMessages`. Messages scheduled to be sent are pushed to this field.
After, if the first range is contiguous to the last sent position, remove the first range and update the last sent position to the range's upper bound.
So, if it is not initialized by `individualDeletedMessages`, then the last sent position can be stuck because of "sent-hole".

More specifically, the details may differ because the definitions are different.

> Should we add metrics to describe how much memory individuallySentPositions usage?

I'll add these metrics to the subscription stats.
```java
    /** The last sent position of the cursor. This is for Key_Shared subscription. */
    public String lastSentPosition;

    /** Set of individually sent ranges. This is for Key_Shared subscription. */
    public String individuallySentPositions;
```

> Should we add a mechanism to stop delivering messages to the client if individuallySentPositions uses more memory than expected?

I referred to the definition of the individualDeletedMessages. It has no limitation to persist on the memory(not [the storage](https://github.com/apache/pulsar/blob/e220a5d04ae16d1b8dfd7e35cdddf43f3a43fe86/pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java#L2049-L2058)).
(Of course, we can add the limitation if necessary.)

Reply:
@equanz 

> I'll add these metrics to the subscription stats.

>    /** The last sent position of the cursor. This is for Key_Shared subscription. */
>    public String lastSentPosition;

>    /** Set of individually sent ranges. This is for Key_Shared subscription. */
>    public String individuallySentPositions;

But `individuallySentPositions` doesn't accurately reflect how much memory `individuallySentPositions` uses, right?

 If there are a huge number of elements in `individuallySentPositions`, it is possible that `pulsar-admin topics stats` might not work, because the response body is too large (we have encountered responses close to 200m).

Reply:
@poorbarcode 

> But individuallySentPositions in topic stats doesn't accurately reflect how much memory individuallySentPositions uses, right?

Yes.
However, I think we can't do effective operations even if we can observe these metrics. What do you think?

> If there are a huge number of elements in individuallySentPositions, it is possible that pulsar-admin topics stats might not work, because the response body is too large (we have encountered responses close to 200m).

Your concerns are correct. I'll reconsider any other approaches(e.g. add a new REST API to expose these stats, just output log as debug level).
BTW, it is not considered in stats(e.g. [consumersAfterMarkDeletePosition](https://github.com/apache/pulsar/blob/e220a5d04ae16d1b8dfd7e35cdddf43f3a43fe86/pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/SubscriptionStatsImpl.java#L126-L127), [keyHashRanges](https://github.com/apache/pulsar/blob/e220a5d04ae16d1b8dfd7e35cdddf43f3a43fe86/pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/stats/ConsumerStatsImpl.java#L96-L97)) and stats-internal(e.g. [individuallyDeletedMessages](https://github.com/apache/pulsar/blob/e220a5d04ae16d1b8dfd7e35cdddf43f3a43fe86/pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/ManagedLedgerInternalStats.java#L94)). We should care about these stats, too.

Reply:
@equanz 

> it is not considered in stats(e.g. consumersAfterMarkDeletePosition, keyHashRanges) and stats-internal(e.g. individuallyDeletedMessages). We should care about these stats, too.

Yes, I'm planning to do it. 
- [ ] compress the `individuallyDeletedMessages` to save place using, see: https://github.com/apache/pulsar/pull/21105
- [ ] A new PIP to limit the place usage of `individuallyDeletedMessages`, the PIP will be submitted today

> Your concerns are correct. I'll reconsider any other approaches(e.g. add a new REST API to expose these stats, just output log as debug level).

After we added new metrics which indicate how much memory is used by `individuallySentPositions`. Push an alert when the memory limit is exceeded, so there are no scenarios that the HTTP API cannot handle

Reply:
> A new PIP to limit the place usage of individuallyDeletedMessages, the PIP will be submitted today

Okay.
If implementation is not far off, I will wait for your new design for reference.
(This PIP issue is one of bug fixes. So, I think we should fix it sooner if possible.)

Reply:
Sure. It is https://github.com/apache/pulsar/pull/21118

Reply:
@poorbarcode 
Sorry for the late reply. I have reconsidered your comments.

This field is appended when messages sent are not contiguous, and is automatically removed when the delivery becomes contiguous.
The frequency of occurrence between instances where messages are not delivered and instances where messages are not acked, would likely differ.
We could add limitations as you mentioned. However, it raises questions about how often this would be effective.

Additionally, introducing such additional stoppage specifications for deliveries would impose more complex limitations on Key_Shared alone.
If these limitations work correctly, that would be ideal. However, past bugs suggest that these complexities often cause omission from consideration.
Hence, if these added limitations do not effectively contribute, it may be better to consider not imposing them in the first place.